File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxHeight = INT_MIN;
    int heightFinder(TreeNode* root){
        if(root==NULL) return 0;
        int lh=heightFinder(root->left);
        int rh=heightFinder(root->right);
        return 1+max(lh,rh);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        
        if(root==NULL) return 0;

        int lh=heightFinder(root->left);
        int rh=heightFinder(root->right);

        maxHeight=max(maxHeight,rh+lh);

        diameterOfBinaryTree(root->left);
        diameterOfBinaryTree(root->right);

        return maxHeight;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    maxi = float('-inf')
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxi = 0

        def dfs(root):
            nonlocal maxi
            if root is None:
                return 0

            l = dfs(root.left)
            r = dfs(root.right)
            maxi = max( maxi , l + r)

            return 1 + max(l,r)

        dfs(root)
        return maxi